src/main/java/com/avispl/symphony/dal/infrastructure/management/tripleplay/TriplePlayAggregatorCommunicator.java

```diff
@@ -0,0 +1,714 @@
+	 * @return String response
+	 * @throws IllegalStateException when exception occur
+	 */
+	private void addAdvanceControlProperties(List<AdvancedControllableProperty> advancedControllableProperties, Map<String, String> stats, AdvancedControllableProperty property) {
+		if (property != null) {
+			for (AdvancedControllableProperty controllableProperty : advancedControllableProperties) {
+				if (controllableProperty.getName().equals(property.getName())) {
+					advancedControllableProperties.remove(controllableProperty);
+					break;
+				}
+			}
+			stats.put(property.getName(), TriplePlayConstrant.EMPTY);
+			advancedControllableProperties.add(property);
+		}
+	}
+
+	/**
+	 * Get key value of aggregated device in cached client list
+	 *
+	 * @param client client to get key value
+	 * @return key of client
+	 */
+	private String getKeyInAggregatedDevice(AggregatedDevice client) {
+		return Optional.ofNullable(client.getProperties().get(NetworkMetric.MAC_ADDRESS.getName())).orElse(null);
+	}
+
+	/**
+	 * Get key value of client in cached client list
+	 *
+	 * @param client client to get key value
+	 * @return key of client
+	 */
+	private String getKeyInCachedClients(Client client) {
+		return Optional.ofNullable(client.getNetwork().getMac()).orElse(null);
+	}
+
+	/**
+	 * Get display name of locale
+	 *
+	 * @param locale locale need to get display name, it must follow format locale in response of tripleplay
+	 */
+	private String getDisplayLocalByLocale(String locale) {
+		if (locale.contains(TriplePlayConstrant.SPLIT_LOCALE)) {
+			Locale local = new Locale(locale.substring(0, locale.lastIndexOf(TriplePlayConstrant.SPLIT_LOCALE)), locale.substring(locale.lastIndexOf(TriplePlayConstrant.SPLIT_LOCALE) + 1));
+			return local.getDisplayName();
+		}
+		return TriplePlayConstrant.EMPTY;
+	}
+
+	/**
+	 * Providing the list of services from cachedServices
+	 */
+	private void convertServiceToAllService(Client client) {
+		client.setServices(cachedServices.values().stream().collect(Collectors.toCollection(ArrayList::new)));
+	}
+
+	/**
+	 * update service in cached clients
+	 *
+	 * @param client client to update
+	 * @param cachedClient client in cached clients list need to update
+	 * @param value name of service to set
+	 */
+	private void updateServiceInCachedClients(AggregatedDevice client, Client cachedClient, String value) {
+		for (AdvancedControllableProperty advancedControllableProperty : client.getControllableProperties()) {
+			if (advancedControllableProperty.getName() == buildNameForChannel(ChannelMetric.SELECT_CHANNEL.getName())) {
+				client.getProperties().put(buildNameForChannel(ChannelMetric.LAST_CHANNEL.getName()), advancedControllableProperty.getValue().toString());
+				break;
+			}
+		}
+		for (Service service : cachedClient.getServices()) {
+			if (service.getName() == client.getProperties().get(buildNameForChannel(ChannelMetric.LAST_CHANNEL.getName()))) {
```

Comment:
service.getName() is also a string, don't you need equals() here and below?

Reply:
I fixed it